external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/metrics/KafkaOffsetMetric.java

```diff
@@ -31,77 +32,75 @@
+
+        Map<TopicPartition, OffsetManager> offsetManagers = offsetManagerSupplier.get();
+        KafkaConsumer kafkaConsumer = consumerSupplier.get();
+
+        if (offsetManagers == null || offsetManagers.isEmpty() || kafkaConsumer == null) {
+            LOG.info("Metrics Tick: offsetManagers or kafkaConsumer is null.");
+            return null;
+        }
+
+        Map<String,TopicMetrics> topicMetricsMap = new HashMap<>();
+        Set<TopicPartition> topicPartitions = offsetManagers.keySet();
+
+        Map<TopicPartition, Long> beginningOffsets= kafkaConsumer.beginningOffsets(topicPartitions);
+        Map<TopicPartition, Long> endOffsets= kafkaConsumer.endOffsets(topicPartitions);
+        Map<String, Long> result = new HashMap<>();
+
+        for (Map.Entry<TopicPartition, OffsetManager> entry : offsetManagers.entrySet()) {
+            TopicPartition topicPartition = entry.getKey();
+            OffsetManager offsetManager = entry.getValue();
+
+            long latestTimeOffset = endOffsets.get(topicPartition);
+            long earliestTimeOffset = beginningOffsets.get(topicPartition);
+
+            long latestEmittedOffset = offsetManager.getLatestEmittedOffset();
+            long latestCompletedOffset = offsetManager.getCommittedOffset();
+            long spoutLag = latestTimeOffset - latestCompletedOffset;
+
+            String metricPath = topicPartition.topic()  + "/partition_" + topicPartition.partition();
+            result.put(metricPath + "/" + "spoutLag", spoutLag);
+            result.put(metricPath + "/" + "earliestTimeOffset", earliestTimeOffset);
+            result.put(metricPath + "/" + "latestTimeOffset", latestTimeOffset);
+            result.put(metricPath + "/" + "latestEmittedOffset", latestEmittedOffset);
+            result.put(metricPath + "/" + "latestCompletedOffset", latestCompletedOffset);
+
+            TopicMetrics topicMetrics = topicMetricsMap.get(topicPartition.topic());
+            if(topicMetrics == null) {
+                topicMetrics = new TopicMetrics();
+                topicMetricsMap.put(topicPartition.topic(), topicMetrics);
             }
-        } catch (Exception t) {
-            LOG.warn("Metrics Tick: Exception when computing kafkaOffset metric.", t);
+
+            topicMetrics.totalSpoutLag += spoutLag;
+            topicMetrics.totalEarliestTimeOffset += earliestTimeOffset;
+            topicMetrics.totalLatestTimeOffset += latestTimeOffset;
+            topicMetrics.totalLatestEmittedOffset += latestEmittedOffset;
+            topicMetrics.totalLatestCompletedOffset += latestCompletedOffset;
+        }
+
+        for(Map.Entry<String, TopicMetrics> e : topicMetricsMap.entrySet()) {
+            String topic = e.getKey();
+            TopicMetrics topicMetrics = e.getValue();
+            result.put(topic + "/" + "totalSpoutLag", topicMetrics.totalSpoutLag);
+            result.put(topic + "/" + "totalEarliestTimeOffset", topicMetrics.totalEarliestTimeOffset);
```

Comment:
I'm not sure I understand a few of these metrics. What does e.g. totalEarliestTimeOffset tell the user? To me it looks like totalSpoutLag makes a lot of sense to have, but the other metric totals aren't very meaningful.

Just to give an example, if I get out a value of 10 for totalLatestTimeOffset, I don't understand what use that could have, because it's just telling me that if I add together the log end offsets of my assigned partitions, I get 10. It could be used to calculate how many messages are in all the partitions, but in that case it makes more sense to return e.g. `totalRecordsInAssignedPartitions = totalLatestTimeOffset - totalEarliestTimeOffset - 1` (disregarding log compaction).

Reply:
@srdo  I agree. totalEarliestTimeOffset alone is not much useful. I just added it to make the metrics compatible with old spout (storm-kafka) metrics. 

adding totalRecordsInAssignedPartitions is good idea. I have included totalRecordsInAssignedPartitions to metrics. Can you explain why we need to decrease the count by 1?  I am not understood what do you mean by disregarding log compaction?

Reply:
The reason I'm asking is because I'd rather only copy over the useful metrics from storm-kafka. It's easier to not add metrics we don't think are useful, than to remove them later.

Sorry, I made a mistake. We don't need to subtract 1. My bad.

Regarding why we disregard log compaction, if there's a partition with a log end offset (latestTimeOffset) of 10, and a beginningOffset of 0, we'd expect there to be 10 messages. When log compaction is enabled, some of the messages between 0 and 10 may have been replaced by tombstones, so they're not there anymore. The metric will still show 10 messages present, because we have no way to know if some of the messages between the beginningOffset and log end offset have been compacted away.

Great, thanks for taking the time to verify that the reflection code works.